### DISCUSSION TITLE
Explanation of __next__ as Field's default_factory

### DISCUSSION THREAD

**Discussion by RGrze** [pydantic/pydantic#3624] [Question]
Hello,

recently I've stumbled upon a weird behavior and I can't find the answer. I'm using Python 3.8 and pydantic 1.8.2

I wanted to make a very simple message counter that is shared across many models and will increment with each instance of the message:

```
from itertools import count
import pydantic

cntr = count()


class BaseM(pydantic.BaseModel):
    mid: int = pydantic.Field(default_factory=cntr.__next__)

class SecondM(BaseM):
    xd: str


print(BaseM())
print(SecondM(xd='xd'))
print(SecondM(xd='xd'))

# output is:
# mid=0
# mid=0 xd='xd'
# mid=1 xd='xd'
```

simple workaround is to just use `lambda: next(cntr)` or `lambda: cntr.__next__` - but why do I need to wrap `__next__` with the lambda expression? Even `def update(): return cntr.__next__()` works.

**Comment by uriyyo** [comment]
That's because of `smart_deepcopy` function that is used by `pydantic`. 

When you define `SecondM` model it will `smart_deepcopy` all fields from the parent model `BaseM`.
`cntr.__next__` will be deepcopied, but `lambda: next(cntr)` and `def update(): return cntr.__next__()` is not.

I will show you examples to make things more clear:
```py
from itertools import count
from pydantic.utils import smart_deepcopy

cntr = count()


def update():
    return next(cntr)


a = cntr.__next__
b = lambda: next(cntr)
c = update

print(smart_deepcopy(a) is a)  # False, will return new object, that's why you received such behavior
print(smart_deepcopy(b) is b)  # True
print(smart_deepcopy(c) is c)  # True
```
